File: tests/csharp/include/ExtendedMessages.cs
/**
 * Extended test message definitions (C#).
 * Three representative extended IDs (boundary/mid/high) + two large-payload
 * messages + five variable-array fill levels = 10 total.
 */

using System;
using System.Text;
using StructFrame;
using StructFrame.ExtendedTest;

namespace StructFrameTests
{
    public static class ExtendedMessages
    {
        public const int MESSAGE_COUNT = 10;

        // ============================================================================
        // Helper functions
        // ============================================================================

        private static ExtendedIdMessage10 CreateExtId10()
        {
            var msg = new ExtendedIdMessage10();
            msg.SmallValue = 256;
            var text = Encoding.UTF8.GetBytes("Boundary Test");
            msg.ShortText = new byte[16];
            Array.Copy(text, msg.ShortText, Math.Min(text.Length, 16));
            msg.Flag = true;
            return msg;
        }

        private static ExtendedIdMessage2 CreateExtId2()
        {
            var msg = new ExtendedIdMessage2();
            msg.SensorId = -42;
            msg.Reading = 2.718281828;
            msg.StatusCode = 50000;
            var desc = "Extended ID test message 2";
            msg.DescriptionLength = (byte)desc.Length;
            msg.DescriptionData = new byte[64];
            Array.Copy(Encoding.UTF8.GetBytes(desc), msg.DescriptionData, desc.Length);
            return msg;
        }

        private static ExtendedIdMessage9 CreateExtId9()
        {
            var msg = new ExtendedIdMessage9();
            msg.BigNumber = -9223372036854775807L;
            msg.BigUnsigned = 18446744073709551615UL;
            msg.PrecisionValue = 1.7976931348623157e+308;
            return msg;
        }

        private static LargePayloadMessage1 CreateLarge1()
        {
            var msg = new LargePayloadMessage1();
            msg.SensorReadings = new float[64];
            for (int i = 0; i < 64; i++) msg.SensorReadings[i] = (float)(i + 1);
            msg.ReadingCount = 64;
            msg.Timestamp = 1704067200000000L;
            var name = Encoding.UTF8.GetBytes("Large Sensor Array Device");
            msg.DeviceName = new byte[32];
            Array.Copy(name, msg.DeviceName, Math.Min(name.Length, 32));
            return msg;
        }

        private static LargePayloadMessage2 CreateLarge2()
        {
            var msg = new LargePayloadMessage2();
            msg.LargeData = new byte[280];
            for (int i = 0; i < 256; i++) msg.LargeData[i] = (byte)i;
            for (int i = 256; i < 280; i++) msg.LargeData[i] = (byte)(i - 256);
            return msg;
        }

        private static ExtendedVariableSingleArray CreateExtVarSingle(ulong ts, byte count, byte[] data, uint crc)
        {
            var msg = new ExtendedVariableSingleArray();
            msg.Timestamp = ts;
            msg.TelemetryDataCount = count;
            msg.TelemetryDataData = new byte[250];
            if (data != null) Array.Copy(data, msg.TelemetryDataData, Math.Min(data.Length, 250));
            msg.Crc = crc;
            return msg;
        }

        // ============================================================================
        // GetMessage(index) - order: ExtId10, ExtId2, ExtId9, Large1, Large2, Var×5
        // ============================================================================

        public static IStructFrameMessage GetMessage(int index)
        {
            switch (index)
            {
                case 0: return CreateExtId10();
                case 1: return CreateExtId2();
                case 2: return CreateExtId9();
                case 3: return CreateLarge1();
                case 4: return CreateLarge2();
                case 5: return CreateExtVarSingle(1, 0,   null,                                          1);
                case 6: return CreateExtVarSingle(2, 1,   new byte[]{ 42 },                              2);
                case 7: { var d = new byte[83];  for (int i=0;i<83;i++)  d[i]=(byte)i;       return CreateExtVarSingle(3, 83,  d, 3); }
                case 8: { var d = new byte[249]; for (int i=0;i<249;i++) d[i]=(byte)(i%256); return CreateExtVarSingle(4, 249, d, 4); }
                default:{ var d = new byte[250]; for (int i=0;i<250;i++) d[i]=(byte)(i%256); return CreateExtVarSingle(5, 250, d, 5); }
            }
        }

        // ============================================================================
        // CheckMessage(index, info) - validates decoded message
        // ============================================================================

        public static bool CheckMessage(int index, FrameMsgInfo info)
        {
            var expected = GetMessage(index);
            if (info.MsgId != expected.GetMsgId()) return false;

            IStructFrameMessage decoded = null;
            if      (info.MsgId == ExtendedIdMessage10.MsgId)           decoded = ExtendedIdMessage10.Deserialize(info);
            else if (info.MsgId == ExtendedIdMessage2.MsgId)            decoded = ExtendedIdMessage2.Deserialize(info);
            else if (info.MsgId == ExtendedIdMessage9.MsgId)            decoded = ExtendedIdMessage9.Deserialize(info);
            else if (info.MsgId == LargePayloadMessage1.MsgId)          decoded = LargePayloadMessage1.Deserialize(info);
            else if (info.MsgId == LargePayloadMessage2.MsgId)          decoded = LargePayloadMessage2.Deserialize(info);
            else if (info.MsgId == ExtendedVariableSingleArray.MsgId)   decoded = ExtendedVariableSingleArray.Deserialize(info);

            if (decoded == null) return false;

            var decodedBytes  = decoded.Serialize();
            var expectedBytes = expected.Serialize();
            if (decodedBytes.Length != expectedBytes.Length) return false;
            for (int i = 0; i < decodedBytes.Length; i++)
                if (decodedBytes[i] != expectedBytes[i]) return false;
            return true;
        }
    }
}

File: tests/js/include/extended_messages.js
/**
 * Extended test message definitions (JavaScript).
 * Three representative extended IDs (boundary/mid/high) + two large-payload
 * messages + five variable-array fill levels = 10 total.
 */

const {
  ExtendedIdMessage2,
  ExtendedIdMessage9,
  ExtendedIdMessage10,
  LargePayloadMessage1,
  LargePayloadMessage2,
  ExtendedVariableSingleArray,
} = require('../../generated/js/extended-test.structframe');

const MESSAGE_COUNT = 10;

// ============================================================================
// Helper functions
// ============================================================================

function createExtId10() {
  return new ExtendedIdMessage10({ smallValue: 256, shortText: 'Boundary Test', flag: true });
}

function createExtId2() {
  return new ExtendedIdMessage2({
    sensorId: -42,
    reading: 2.718281828,
    statusCode: 50000,
    descriptionLength: 'Extended ID test message 2'.length,
    descriptionData: 'Extended ID test message 2',
  });
}

function createExtId9() {
  return new ExtendedIdMessage9({
    bigNumber: -9223372036854775807n,
    bigUnsigned: 18446744073709551615n,
    precisionValue: 1.7976931348623157e+308,
  });
}

function createLarge1() {
  return new LargePayloadMessage1({
    sensorReadings: Array.from({ length: 64 }, (_, i) => i + 1),
    readingCount: 64,
    timestamp: 1704067200000000n,
    deviceName: 'Large Sensor Array Device',
  });
}

function createLarge2() {
  const largeData = [];
  for (let i = 0; i < 256; i++) largeData.push(i);
  for (let i = 256; i < 280; i++) largeData.push(i - 256);
  return new LargePayloadMessage2({ largeData });
}

function extVar(ts, count, data, crc) {
  return new ExtendedVariableSingleArray({
    timestamp: ts,
    telemetryDataCount: count,
    telemetryDataData: data,
    crc,
  });
}

// ============================================================================
// getMessage(index) - order: ExtId10, ExtId2, ExtId9, Large1, Large2, Var×5
// ============================================================================

function getMessage(index) {
  switch (index) {
    case 0: return createExtId10();
    case 1: return createExtId2();
    case 2: return createExtId9();
    case 3: return createLarge1();
    case 4: return createLarge2();
    case 5: return extVar(0x0000000000000001n, 0,   [],                                            0x00000001);
    case 6: return extVar(0x0000000000000002n, 1,   [42],                                          0x00000002);
    case 7: return extVar(0x0000000000000003n, 83,  Array.from({ length: 83  }, (_, i) => i),      0x00000003);
    case 8: return extVar(0x0000000000000004n, 249, Array.from({ length: 249 }, (_, i) => i % 256), 0x00000004);
    default: return extVar(0x0000000000000005n, 250, Array.from({ length: 250 }, (_, i) => i % 256), 0x00000005);
  }
}

// ============================================================================
// checkMessage(index, info) - validates decoded message
// ============================================================================

function checkMessage(index, info) {
  const expected = getMessage(index);
  const msgClass = expected.constructor;
  if (info.msgId !== msgClass._msgid) return false;
  const decoded = msgClass.deserialize(info);
  return decoded.equals(expected);
}

module.exports = { MESSAGE_COUNT, getMessage, checkMessage };
